File: AoC_2024/AoC_2024_01/Main.hs
----------------------------------------------------
--              Advent of Code 2024               --
--           Day 1: Historian Hysteria            --
--            Solution by Lorin Lange             --
----------------------------------------------------

module Main where

import Data.List ( sort, group )
import Data.Maybe ( fromMaybe )

makeScore :: [(Int, Int)] -> Int -> Int
makeScore scores n = fromMaybe 0 (lookup n scores) * n

parse :: String -> (Int, Int)
parse str = (read $ head lst, read $ lst !! 1)
    where lst = words str

main :: IO()
main = do
    lst <- map parse . lines <$> readFile "input.txt"
    let lst1 = sort $ map fst lst
    let lst2 = sort $ map snd lst

    putStrLn $ "Part 1: " ++ show (sum $ zipWith (\a b -> abs $ a - b) lst1 lst2)

    let freq = map (\l -> (head l, length l)) $ group lst2
    putStrLn $ "Part 2: " ++ show (sum $ map (makeScore freq) lst1)

File: AoC_2024/AoC_2024_01/main.rs
////////////////////////////////////////////////////
//              Advent of Code 2024               //
//           Day 1: Historian Hysteria            //
//            Solution by Lorin Lange             //
////////////////////////////////////////////////////

use std::fs;
use std::collections::HashMap;

fn parse(path : &str) -> (Vec<i32>, Vec<i32>) {
    let input = fs::read_to_string(path).expect("Couldn't read file.");
    let lines = input.lines().collect::<Vec<_>>();
    let (mut l1, mut l2) = (Vec::new(), Vec::new());
    for line in lines {
        let mut st = line.split_whitespace();
        let mut n = st.next().unwrap().parse().unwrap();
        l1.push(n);
        n = st.next().unwrap().parse().unwrap();
        l2.push(n);
    }
    l1.sort();
    l2.sort();
    (l1, l2)
}

fn part_1(l1 : &Vec<i32>, l2 : &Vec<i32>) -> i32 {
    l1.iter().zip(l2).map(|(x, y)| (x - y).abs()).sum()
}

fn part_2(l1 : &Vec<i32>, l2 : &Vec<i32>) -> i32 {
    let mut map = HashMap::new();
    for &x in l2 {
        *map.entry(x).or_insert(0) += 1;
    }
    l1.iter().map(|&a| a * map.get(&a).unwrap_or(&0)).sum()
}

#[test]
fn test_input() {
    let (lst1, lst2) = parse("test_input.txt");
    assert_eq!(part_1(&lst1, &lst2), 11);
    assert_eq!(part_2(&lst1, &lst2), 31);
}

#[test]
fn input() {
    let (lst1, lst2) = parse("input.txt");
    assert_eq!(part_1(&lst1, &lst2), 2430334);
    assert_eq!(part_2(&lst1, &lst2), 28786472);
}

fn main() {
    let (lst1, lst2) = parse("input.txt");
    println!("Part 1: {}", part_1(&lst1, &lst2));
    println!("Part 2: {}", part_2(&lst1, &lst2));
}
